### PR TITLE
feat(payments): add Email Verification to Passwordless Flow

### PR DESCRIPTION
**Pull request by IvoJP**
Because:

* We want to prevent subscriptions created for fraudulent purposes that use invalid email domains

This commit:

* As a temp solution refactors the email domain check used by content server and piggy-backs off the account status check to return information about the validity of the email domain

Closes https://github.com/mozilla/fxa/issues/12406
## Checklist

_Put an `x` in the boxes that apply_

- [X] My commit is GPG signed.
- [X] If applicable, I have modified or added tests which pass locally.
- [X] I have added necessary documentation (if appropriate).
- [X] I have verified that my changes render correctly in RTL (if appropriate).

## Other information (Optional)
This is a temporary solution until we evaluate alternatives in #12405 and determine what a more permanent solution will be. I wanted this to be minimally invasive in how things currently work and as easily to remove/revert as possible later.

### CODE REVIEW

File: packages/fxa-shared/email/emailValidatorErrors.js [hunk 1]
```diff
@@ -0,0 +1,28 @@
+const dns = require('dns');
```
Comment (bbangert) [inline packages/fxa-shared/email/emailValidatorErrors.js:1]: New files, wherever possible, should be TypeScript. Was it not compiling properly in this location when using a `.ts` file?

File: packages/fxa-auth-server/lib/routes/account.ts [hunk 2]
```diff
@@ -1027,17 +1029,49 @@ export class AccountHandler {
...
+      const domain = email.split('@')[1];
+      if (!TopEmailDomains.has(domain)) {
+        let mxCheck = false,
+          ipv4Check = false;
+
+        try {
+          mxCheck = await tryResolveMx(domain);
+          ipv4Check = await tryResolveIpv4(domain);
+        } catch (error) {
+          // check against email errors
```
Comment (bbangert) [inline packages/fxa-auth-server/lib/routes/account.ts:1045]: If we error, we silently ignore it?

File: packages/fxa-shared/email/validateEmail.js [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+export const results = ['MX', 'A', 'none', 'skip'].reduce(
+  (accumulator, val) => {
+    accumulator[val] = val;
+    return accumulator;
+  },
+  {}
+);
+
+const tryResolveWith = (resolveFunc) => async (domain) => {
```
Comment (bbangert) [inline packages/fxa-shared/email/validateEmail.js:13]: Why not write this as a standard function?

File: packages/fxa-shared/email/validateEmail.js [hunk 4]
```diff
@@ -0,0 +1,32 @@
+const dns = require('dns');
+const resolver = new dns.promises.Resolver();
+const NotFoundErrorCodes = require('./emailValidatorErrors').NotFoundErrorCodes;
```
Comment (bbangert) [inline packages/fxa-shared/email/validateEmail.js:3]: Mixing up variable declarations with imports doesn't follow our style guidelines (which are poorly defined) or existing code conventions.

File: packages/fxa-shared/email/validateEmail.js [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+export const results = ['MX', 'A', 'none', 'skip'].reduce(
```
Comment (bbangert) [inline packages/fxa-shared/email/validateEmail.js:5]: What kind of results are these? This is pretty hard to follow.

File: packages/fxa-shared/email/validateEmail.js [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+  } catch (err) {
+    if (NotFoundErrorCodes.includes(err.code)) {
+      return false;
+    }
+    throw err;
+  }
+};
+
+export const tryResolveMx = async (domain) => {
+  return await tryResolveWith(resolver.resolveMx.bind(resolver))(domain);
```
Comment (bbangert) [inline packages/fxa-shared/email/validateEmail.js:27]: Why are these being awaited here?

**Review by bbangert — CHANGES_REQUESTED** [review]

**Comment by bbangert** [comment]
This seems to only hook the email check into a client-driven API call that it checks. If a user tweaks their local code to ignore the invalid response here, will our other account creation code run this email domain check as well? If not, they can still bypass this.

File: packages/fxa-shared/email/validateEmail.js (hunk 3 above)
Comment (IvoJP) [reply packages/fxa-shared/email/validateEmail.js:13]: this is just a refactor of the existing check as a temporary solution. this is how the function existed in `fxa-content-server`

i agree with your sentiment that this is poorly written but I am hesitant to rework any of the logic here rather than just pulling it out to `fxa-shared`

**Comment by IvoJP** [comment]
@StaberindeZA not sure why content errors on it, i will look into it but `jj.com` is valid - its johnson and johnson's website so the payments server is correct in behavior in what your described but i could have done something that affected the content server from handling correctly.

File: packages/fxa-shared/email/emailValidatorErrors.js (hunk 1 above)
Comment (IvoJP) [reply packages/fxa-shared/email/emailValidatorErrors.js]: updated

File: packages/fxa-shared/email/validateEmail.js (hunk 6 above)
Comment (IvoJP) [reply packages/fxa-shared/email/validateEmail.js]: confirmed, unnecessary and left over from developement. removed.

**Comment by StaberindeZA** [comment]
> i will look into it but `jj.com` is valid

I had a quick look, and according to mailbox.validator.com the `jj.com` domain has no active MX records, and I believe I saw something in the code checking for that? Might explain why the error pops up in production.

image

**Comment by IvoJP** [comment]
Yeah I think it is because the content server warns if it can resolve the domain to an IP address but not find a MX record for the domain. If thats the case, the content server actually lets you proceed, you can click "Continue" again and it will dismiss the warning and proceed, whereas if it can't resolve either a MX record or IP address it blocks it entirely.

What I am missing, i guess, from the payments server is the warning but I am not sure with how we have the payments server setup if i could do something similar. I'll take a look but my vote would be to ignore the warning since we do allow users to use a domain that resolves to an IP address even if an MX record cannot be found.

File: packages/fxa-shared/email/validateEmail.js (hunk 5 above)
Comment (IvoJP) [reply packages/fxa-shared/email/validateEmail.js]: since i dont need this in the auth server i am moving it back to where it came from and leaving the original logic as is

File: packages/fxa-auth-server/lib/routes/account.ts (hunk 2 above)
Comment (IvoJP) [reply packages/fxa-auth-server/lib/routes/account.ts]: I added some logging around it. The behavior as it currently exists in the content server is that if any error happens during this check, we let the user proceed without checking if the domain is valid. Let me know if the logging I added is sufficient.

File: packages/fxa-shared/email/validateEmail.js (hunk 3 above)
Comment (bbangert) [reply packages/fxa-shared/email/validateEmail.js]: Ah, got it. These two files should have the MPL added to the top.

File: packages/fxa-shared/email/validateEmail.js (hunk 4 above)
Comment (IvoJP) [reply packages/fxa-shared/email/validateEmail.js]: I have updated this to `import { NotFoundErrorCodes } from './emailValidatorErrors';` let me know if that rectifies  your issue with this line.

just a heads-up (also to clarify in case I am misunderstanding this) we do have the following in our code:
```
const release = require('../../package.json').version;
const config = require('../config').getProperties();
const METRICS_CONTEXT_SCHEMA = require('../metrics/context').schema;
const jwtverify = require('util').promisify(jsonwebtoken.verify);
const hex = require('buf').to.hex;
```
i am not sure if these examples are similar or different enough, so pardon my ignorance.

i've brought up this issue with anna before, is there a style guide you can reference?  it seems our we often have mixed conventions and then it comes during code review that something is against convention when our code base isnt _seemingly_ consistent enough to tease this out during development time. 

let me know if the other examples i posted arent what you meant, i am trying to learn how to avoid these mistakes going forward.

File: packages/fxa-shared/email/validateEmail.ts [hunk 7]
```diff
@@ -0,0 +1,25 @@
+import * as dns from 'dns';
```
Comment (bbangert) [inline packages/fxa-shared/email/validateEmail.ts:1]: Need license string above.

File: packages/fxa-shared/email/emailValidatorErrors.ts [hunk 8]
```diff
@@ -0,0 +1,28 @@
+const dns = require('dns');
```
Comment (bbangert) [inline packages/fxa-shared/email/emailValidatorErrors.ts:1]: Needs license string above.

File: packages/fxa-shared/email/validateEmail.ts (hunk 7 above)
Comment (IvoJP) [reply packages/fxa-shared/email/validateEmail.ts:4]: done

File: packages/fxa-shared/email/validateEmail.js (hunk 3 above)
Comment (IvoJP) [reply packages/fxa-shared/email/validateEmail.js]: added

File: packages/fxa-payments-server/public/locales/en-US/main.ftl [hunk 9]
```diff
@@ -427,6 +427,7 @@ new-user-subscribe-product-assurance = We only use your email to create your acc
 new-user-email-validate = Email is not valid
 new-user-email-validate-confirm = Emails do not match
 new-user-already-has-account-sign-in = You already have an account. <a>Sign in</a>
+new-user-invalid-email-domain = Mistyped email? { $domain } does not offer email.
```
[the remark this answers is not in the document]
Comment (IvoJP) [reply packages/fxa-payments-server/public/locales/en-US/main.ftl:431]: added

**Comment by IvoJP** [comment]
@bbangert I caught an error in domain check so i added basically a bypess if there is any error; in the error check now it sets both checks (mail exchange and ipv4) to true so that we don't block anyone from signing up if domain checker fails for any reason. 

To answer your question in your other comment: this check does not happen on account creation, this is purely a FE check. I can look into adding to the account creation as well - my original intent was for this to behave exactly as it does for the content server but obvisously the abuse/fraud is happening on subplat so if we are concerned malicious actors will bypass the check then that necessitates adding it on account creation as well.

@StaberindeZA i briefly looked over the component where we display the error and the app running on the content server - the app on the content server has all 4 scenarios accounted for: skipped, passes, fails mail exchange but not IP (which results in a warning tool tip but does not block the user from moving forward), or fails both (user is blocked from proceeding). I dont think `NewUserEmailForm` is set up for warnings, we just have errors which block the form. Functionally, they allow and permit the same things but on the checkout page we wont have the warning about `Mistyped email?` for domains that dont have an MX record. Let me know if this is blocking for you.

**Review by StaberindeZA — APPROVED** [review]
r+ wc

As you mentioned, this blocks the invalid domains, the same way as the content server. For the "warning" message, could you file a follow up ticket so we can discuss whether or not we should pick it up?

File: packages/fxa-payments-server/src/components/NewUserEmailForm/index.test.tsx [hunk 10]
```diff
@@ -51,9 +54,13 @@ const WrapNewUserEmailForm = ({
         setEmailsMatch={setEmailsMatch}
         setValidEmail={setValidEmail}
         setAccountExists={setAccountExists}
+        setInvalidEmailDomain={setInvalidEmailDomain}
         getString={(id: string) => id}
         checkAccountExists={() =>
-          Promise.resolve({ exists: accountExistsReturnValue })
+          Promise.resolve({
+            exists: accountExistsReturnValue,
+            invalidDomain: invalidDomain,
```
[the remark this answers is not in the document]
Comment (IvoJP) [reply packages/fxa-payments-server/src/components/NewUserEmailForm/index.test.tsx]: done

File: packages/fxa-payments-server/src/components/NewUserEmailForm/index.tsx [hunk 11]
```diff
@@ -184,15 +191,19 @@ export async function emailInputValidationAndAccountCheck(
   focused: boolean,
   setValidEmail: (value: string) => void,
   setAccountExists: (value: boolean) => void,
+  setInvalidEmailDomain: (value: boolean) => void,
   setEmailInputState: (value: string) => void,
-  checkAccountExists: (email: string) => Promise<{ exists: boolean }>,
+  checkAccountExists: (
+    email: string
+  ) => Promise<{ exists: boolean; invalidDomain: boolean | undefined }>,
```
[the remark this answers is not in the document]
Comment (IvoJP) [reply packages/fxa-payments-server/src/components/NewUserEmailForm/index.tsx]: it is not; fixed

File: packages/fxa-payments-server/public/locales/en-US/main.ftl [hunk 12]
```diff
@@ -427,6 +427,8 @@ new-user-subscribe-product-assurance = We only use your email to create your acc
 new-user-email-validate = Email is not valid
 new-user-email-validate-confirm = Emails do not match
 new-user-already-has-account-sign-in = You already have an account. <a>Sign in</a>
+# $domain - the email domain provided by the user during sign up
```
[the remark this answers is not in the document]
Comment (IvoJP) [reply packages/fxa-payments-server/public/locales/en-US/main.ftl]: done

**Comment by IvoJP** [comment]
@bbangert I have added the check to the stub account creation and added tests.

**Review by bbangert — APPROVED** [review]
